azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_mqtt.py

```diff
@@ -41,8 +41,9 @@ def _run_op(self, op):
                 username=self.username,
                 ca_cert=self.ca_cert,
             )
-            self.transport.on_mqtt_connected = self.on_connected
-            self.transport.on_mqtt_disconnected = self.on_disconnected
+            self.transport.on_mqtt_connected = self._on_unexpected_connection
+            self.transport.on_mqtt_connection_failure = self._on_unexpected_connection_failed
+            self.transport.on_mqtt_disconnected = self._on_unexpected_disconnection
```

Comment:
Another question where I think this makes sense.  these handlers (_on_unexpected_connection, etc) are only for cases where Paho calls the handler but we're not trying to connect, so I've added the "unexpected" name to indicate this.  Other cases, like we get connected and we were trying to connect, are handled by the temporary handlers below.  This seems to be the best way to handle this scenario, but it's all ugly so I'm wondering if there's a better way..

Reply:
I see the value in this, but I think the code is somewhat confusing. Perhaps we could find a better way to contain this logic.

Also does the dynamic switching of handlers present any potential threading issues / race conditions?

---
In reply to: [302270134](https://github.com/Azure/azure-iot-sdk-python-preview/pull/136#discussion_r302270134) [](ancestors = 302270134)

Reply:
fixed in my new iteration.

---
In reply to: [303138886](https://github.com/Azure/azure-iot-sdk-python-preview/pull/136#discussion_r303138886) [](ancestors = 303138886,302270134)